ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -216,6 +241,162 @@ KEXT_STATIC void UseMainForkIfNamedStream(
+        }
+        else
+        {
+            arrayToFree = newArray;
+            arrayToFreeLength = newMaxPendingRenames;
+        }
+    }
+    SpinLock_Release(s_renameLock);
+    
+    if (arrayToFree != nullptr)
+    {
+        assert(arrayToFreeLength > 0);
+        Memory_FreeArray(arrayToFree, arrayToFreeLength);
+    }
+}
+
+KEXT_STATIC void RecordPendingRenameOperation(vnode_t vnode)
+{
+    assertf(version_major >= 18, "This function should only be called from the KAUTH_FILEOP_WILL_RENAME handler, which is only supported by Darwin 18 and newer (version_major = %u)", version_major);
+    thread_t myThread = current_thread();
+    
+    bool resizeTable;
+    do
+    {
+        resizeTable = false;
+        uint32_t resizeTableLength = 0;
+        
+        SpinLock_Acquire(s_renameLock);
+        {
+            if (s_pendingRenameCount < s_maxPendingRenames)
+            {
+                s_pendingRenames[s_pendingRenameCount].thread = myThread;
+                s_pendingRenames[s_pendingRenameCount].vnode = vnode;
+                ++s_pendingRenameCount;
+            }
+            else
+            {
+                for (uint32_t i = 0; i < s_pendingRenameCount; ++i)
+                {
+                    assert(s_pendingRenames[i].thread != myThread);
+                }
+
+                resizeTable = true;
+                
+                resizeTableLength = MAX(1u, MIN(UINT32_MAX, s_maxPendingRenames * UINT64_C(2)));
+                assert(resizeTableLength > s_maxPendingRenames);
+            }
+        }
+        SpinLock_Release(s_renameLock);
+        
+        if (resizeTable)
+        {
+            if (resizeTableLength > 16)
+            {
+                KextLog_Error("Warning: RecordPendingRenameOperation is causing pending rename array resize to %u items.", resizeTableLength);
+            }
+            ResizePendingRenames(resizeTableLength);
+        }
+    } while (resizeTable);
+}
+
+KEXT_STATIC bool DeleteOpIsForRename(vnode_t vnode)
+{
+    if (version_major < 18)
+    {
+        // High Sierra and earlier do not support WILL_RENAME notification, so we have to assume any delete may be caused by a rename
+        assert(s_pendingRenameCount == 0);
+        assert(s_maxPendingRenames == 0);
+        return true;
+    }
+    
+    bool isRename = false;
+    
+    thread_t myThread = current_thread();
+    
+    SpinLock_Acquire(s_renameLock);
+    {
+        for (uint32_t i = 0; i < s_pendingRenameCount; ++i)
+        {
+            if (s_pendingRenames[i].thread == myThread)
+            {
+                isRename = true;
+                assert(s_pendingRenames[i].vnode == vnode);
+                --s_pendingRenameCount;
+                if (i != s_pendingRenameCount)
+                {
+                    s_pendingRenames[i] = s_pendingRenames[s_pendingRenameCount];
+                }
+                
+                s_pendingRenames[s_pendingRenameCount] = PendingRenameOperation{};
```

Comment:
No, `current_thread()` never returns `nullptr`. Additionally, because we only check the first `s_pendingRenameCount` array elements, we will never test against a default-initialised entry.